byzcoin/struct.go

```diff
@@ -740,3 +741,75 @@ const (
 	// instances. If not present, missing instances are ignored.
 	GUFSendMissingProofs
 )
+
+// runSingleWG allows to ensure a single execution of a method.
+type runSingleWG struct {
+	sync.WaitGroup
+	running int32
+}
+
+// Increment on Start - returns false if wg is already running
+func (cg *runSingleWG) Start() bool {
+	fresh := atomic.CompareAndSwapInt32(&cg.running, 0, 1)
+	if !fresh {
+		return false
+	}
+	cg.Add(1)
+	return true
+}
+
+// Decrement on Done
+func (cg *runSingleWG) Done() {
+	if !cg.Running() {
+		panic("cannot call done on idle runSingleWG")
+	}
+	atomic.AddInt32(&cg.running, -1)
+	cg.WaitGroup.Done()
+}
+
+// Check if currently running
+func (cg *runSingleWG) Running() bool {
+	return atomic.LoadInt32(&cg.running) > 0
+}
+
+// tasksWG is a special workGroup that can be paused or resumed.
+type tasksWG struct {
```

Comment:
You do not implement `Done` so `tasksWG` can be still running even if all expected children calls `Done`.

Reply:
OK - probably I need to rename `running` here. It is normal for the structure to have `running = true`, but no tasks are waiting.

I can rename `(r|R)unning` to `(t|T)asksAllowed` - would that make it clearer? Other proposition for the name?

Reply:
renamed and added a comment to the structure. Also lowercased all methods on the structure, except for `Add`, as this one overwrites the one from `sync.WaitGroup`.